chore; Update README with comprehensive documentation and badges (#17)
TCatshoek/fastapi-nextauth-jwt: README.md

# fastapi-nextauth-jwt

[![PyPI version](https://badge.fury.io/py/fastapi-nextauth-jwt.svg)](https://badge.fury.io/py/fastapi-nextauth-jwt)
[![PyPI Downloads](https://img.shields.io/pypi/dm/fastapi-nextauth-jwt)](https://pypi.org/project/fastapi-nextauth-jwt/)
[![License](https://img.shields.io/pypi/l/fastapi-nextauth-jwt)](https://github.com/yourusername/fastapi-nextauth-jwt/blob/main/LICENSE)
[![Contributors](https://img.shields.io/github/contributors/TCatshoek/fastapi-nextauth-jwt)](https://github.com/TCatshoek/fastapi-nextauth-jwt/graphs/contributors)

This project provides a FastAPI dependency for decrypting and validating JWTs generated by Auth.js. It is designed to facilitate the integration of a FastAPI backend with Next.js and NextAuth/Auth.js on the frontend.

> [!NOTE]
> Using Auth.js with frameworks other than Next.js may work but has not been tested

> [!IMPORTANT]

> > From version 2.0.0, and the default cookie names have changed (like authjs.session-token). You'll need to update your cookie name configuration accordingly if you're using the latest version of Auth.js.

Reviewer: Sorry if that was unclear, but this part can be removed entirely. The default cookie name is correct in 2.0.0, and nothing needs to be configured manually

# fastapi-nextauth-jwt

[![PyPI version](https://badge.fury.io/py/fastapi-nextauth-jwt.svg)](https://badge.fury.io/py/fastapi-nextauth-jwt)
[![PyPI Downloads](https://img.shields.io/pypi/dm/fastapi-nextauth-jwt)](https://pypi.org/project/fastapi-nextauth-jwt/)
[![License](https://img.shields.io/pypi/l/fastapi-nextauth-jwt)](https://github.com/yourusername/fastapi-nextauth-jwt/blob/main/LICENSE)
[![Contributors](https://img.shields.io/github/contributors/TCatshoek/fastapi-nextauth-jwt)](https://github.com/TCatshoek/fastapi-nextauth-jwt/graphs/contributors)

This project provides a FastAPI dependency for decrypting and validating JWTs generated by Auth.js. It is designed to facilitate the integration of a FastAPI backend with Next.js and NextAuth/Auth.js on the frontend.

> > [!WARNING]

Reviewer: I think this could be !NOTE instead, as it is not really a warning imo

Reviewer: I would probably also word it as something like: Using Auth.js with frameworks other than Next.js may work but has not been tested

> It may also be compatible with other frameworks using Auth.js, though this has not been extensively tested.

> [!IMPORTANT]

> > NextAuth.js has been renamed to Auth.js, and the default cookie names have changed to reflect this (e.g., from `next-auth.csrf-token` to `authjs.csrf-token`). You'll need to update your cookie name configuration accordingly if you're using the latest version of Auth.js.

Reviewer: This is incorrect, starting from version 2.0.0 the default cookie names take this change into account, see: https://github.com/TCatshoek/fastapi-nextauth-jwt/blob/bef787beff050a4c62a96b721bfbc4d8793d3f33/src/fastapi_nextauth_jwt/fastapi_nextauth_jwt.py#L88

## Features

- **JWT Decryption & Validation**: Seamlessly decrypt and validate JWTs generated by Auth.js
- **CSRF Protection**: Built-in Auth.js-compatible CSRF protection with configurable HTTP methods

> - **Secure by Default**: Automatic secure cookie configuration based on your environment

Reviewer: I would prefer to not make any security claims as none of this code has been professionally audited

app = FastAPI()

JWT = NextAuthJWT(
    secret="y0uR_SuP3r_s3cr37_$3cr3t", # Leave this out to automatically read the NEXTAUTH_SECRET env var
)

@app.get("/")
async def return_jwt(jwt: Annotated[dict, Depends(JWT)]):
    return jwt
```

## Configuration Options

### Essential Settings

- **secret** (str): The secret key used for JWT operations. Should match `NEXTAUTH_SECRET` in your Next.js app. Leave this out to automatically read the `NEXTAUTH_SECRET` environment variable.
  ```python
  JWT = NextAuthJWT(secret=os.getenv("YOUR_SECRET_ENV_VAR_NAME")))
  ```

> ### Security Options
> 
> - **csrf_prevention_enabled** (bool): Enable CSRF protection
>   - Defaults to `False` in development (`ENV=dev`), `True` otherwise
> 
> - **csrf_methods** (Set[str]): HTTP methods requiring CSRF protection
>   - Default: `{'POST', 'PUT', 'PATCH', 'DELETE'}`
> 
> ### Cookie Configuration
> 
> - **secure_cookie** (bool): Enable secure cookie attributes
>   - Default: `True` (when `NEXTAUTH_URL` starts with https)
> 
> - **cookie_name** (str): Session token cookie name
>   - Default: `"__Secure-authjs.session-token"` (when secure_cookie is True)
>   - Default: `"authjs.session-token"` (when secure_cookie is False)
> 
> - **csrf_cookie_name** (str): CSRF token cookie name
>   - Default: `"__Host-authjs.csrf-token"` (when secure_cookie is True)
>   - Default: `"authjs.csrf-token"` (when secure_cookie is False)
> 
> > [!TIP]
> > If you're using the latest version of Auth.js, here's the recommended configuration:
> > ```python
> > JWT = NextAuthJWT(
> >     secret=os.environ["AUTHJS_SECRET"],
> > )
> > ```
> 
> ### Advanced Options

Reviewer: Maybe it would be nice to make it clear that you won't need to touch these settings if you use the default Auth.js settings:
```suggestion
### Additional Options
If your auth.js settings are left at their defaults, you shouldn't need to touch these.

#### Security Options

- **csrf_prevention_enabled** (bool): Enable CSRF protection
  - Defaults to `False` in development (`ENV=dev`), `True` otherwise

- **csrf_methods** (Set[str]): HTTP methods requiring CSRF protection
  - Default: `{'POST', 'PUT', 'PATCH', 'DELETE'}`

#### Cookie Configuration

- **secure_cookie** (bool): Enable secure cookie attributes
  - Default: `True` (when `NEXTAUTH_URL` starts with https)

- **cookie_name** (str): Session token cookie name
  - Default: `"__Secure-authjs.session-token"` (when secure_cookie is True)
  - Default: `"authjs.session-token"` (when secure_cookie is False)

- **csrf_cookie_name** (str): CSRF token cookie name
  - Default: `"__Host-authjs.csrf-token"` (when secure_cookie is True)
  - Default: `"authjs.csrf-token"` (when secure_cookie is False)

> [!TIP]
> If you're using the latest version of Auth.js, here's the recommended configuration:
> ```python
> JWT = NextAuthJWT(
>     secret=os.environ["AUTHJS_SECRET"],
> )
> ```

#### Advanced Options
```

- **encryption_algorithm** (str): JWT encryption algorithm
  - Supported: `"A256CBC-HS512"` (default), `"A256GCM"`

- **check_expiry** (bool): Enable JWT expiration validation
  - Default: `True`

## NextAuth.js v4 Compatibility

For NextAuth.js v4 applications, use the `NextAuthJWTv4` class:

```python
from fastapi_nextauth_jwt import NextAuthJWTv4

JWT = NextAuthJWTv4(
    secret=os.getenv("NEXTAUTH_SECRET")
)
```

This provides compatibility with the v4 token format and default settings

> ## Security Best Practices

Reviewer: ```suggestion
## Best Practices
```

app = FastAPI()

JWT = NextAuthJWT(
    secret="y0uR_SuP3r_s3cr37_$3cr3t", # Leave this out to automatically read the NEXTAUTH_SECRET env var
)

@app.get("/")
async def return_jwt(jwt: Annotated[dict, Depends(JWT)]):
    return jwt
```

## Configuration Options

### Essential Settings

> - **secret** (str): The secret key used for JWT operations. Should match `NEXTAUTH_SECRET` in your Next.js app.
>   ```python
>   JWT = NextAuthJWT(secret=os.getenv("NEXTAUTH_SECRET"))
>   ```

Reviewer: Since this is read from NEXTAUTH_SECRET by default, maybe this is more clear
```suggestion
- **secret** (str): The secret key used for JWT operations. Should match `NEXTAUTH_SECRET` in your Next.js app. Leave this out to automatically read the `NEXTAUTH_SECRET` environment variable.
  ```python
  JWT = NextAuthJWT(secret=os.getenv("YOUR_SECRET_ENV_VAR_NAME")))
  ```
```

  - Default: `"next-auth.session-token"` (when secure_cookie is False)
  - For Auth.js (latest), use: `"__Secure-authjs.session-token"` (when secure_cookie is True)

- **csrf_cookie_name** (str): CSRF token cookie name
  - Default: `"__Host-next-auth.csrf-token"` (when secure_cookie is True)
  - Default: `"next-auth.csrf-token"` (when secure_cookie is False)
> [!TIP]
> If you're using the latest version of Auth.js, here's the recommended configuration:
> ```python
> JWT = NextAuthJWT(
>     secret=os.environ["AUTHJS_SECRET"],
> )
> ```

### Advanced Options

- **encryption_algorithm** (str): JWT encryption algorithm
  - Supported: `"A256CBC-HS512"` (default), `"A256GCM"`

- **check_expiry** (bool): Enable JWT expiration validation
  - Default: `True`

## NextAuth.js v4 Compatibility

For NextAuth.js v4 applications, use the `NextAuthJWTv4` class:

```python
from fastapi_nextauth_jwt import NextAuthJWTv4

JWT = NextAuthJWTv4(
    secret=os.getenv("NEXTAUTH_SECRET")
)
```

This provides compatibility with the v4 token format and default settings, including:
- Uses `"A256GCM"` encryption algorithm
- Sets specific key derivation parameters
- Uses v4-compatible cookie names

> - Disables salt auto-appending

Reviewer: I feel like these details may be a bit too much for someone just using the library, they may not be interested in the inner workings and it could detract from the information they need to know

```suggestion
This provides compatibility with the v4 token format and default settings
```

## Security Best Practices

1. **Environment Variables**: Always use environment variables for sensitive values:
   ```python
   JWT = NextAuthJWT(
       secret=os.getenv("NEXTAUTH_SECRET"),
   )
   ```

2. **HTTPS in Production**: Ensure `NEXTAUTH_URL` starts with `https://` in production to enable secure cookies

3. **CSRF Protection**: Keep CSRF protection enabled in production environments

## Examples

A [simple example](https://github.com/TCatshoek/fastapi-nextauth-jwt/tree/main/examples/simple) is available in the examples folder. It demonstrates:
- Using Next.js URL rewrites to route requests to FastAPI
- Basic JWT validation setup
- CSRF protection configuration

> You can also place both the backend and frontend behind a reverse proxy like nginx, as long as cookies reach FastAPI.

Reviewer: ```suggestion
You can also place both the backend and frontend behind a reverse proxy like nginx, as long as the auth.js cookies reach FastAPI.
```

app = FastAPI()

JWT = NextAuthJWT(

>     secret="y0uR_SuP3r_s3cr37_$3cr3t",  # Better to use environment variable

Reviewer: ```suggestion
    secret="y0uR_SuP3r_s3cr37_$3cr3t",  # Leave this out to automatically read the NEXTAUTH_SECRET env var
```

)

@app.get("/")
async def return_jwt_get(jwt: Annotated[dict, Depends(JWT)]):
    return {"message": f"Hi {jwt['name']}. Greetings from fastapi!"}

# For CSRF protection testing
@app.post("/")
async def return_jwt_post(jwt: Annotated[dict, Depends(JWT)]):

>     return {"message": f"Hi {jwt['name']}. Greetings from fastapi!"}

Reviewer: Adding this to the example doesn't really add anything imo

```

## Configuration Options

### Essential Settings

- **secret** (str): The secret key used for JWT operations. Should match `NEXTAUTH_SECRET` in your Next.js app.
  ```python
  JWT = NextAuthJWT(secret=os.getenv("NEXTAUTH_SECRET"))
  ```

### Security Options

> - **csrf_prevention_enabled** (bool): Toggle CSRF protection
>   - Defaults to `True` in development (`ENV=dev`), `False` otherwise

Reviewer: Sorry, this is incorrect in the docstring. It should be: 

```suggestion
- **csrf_prevention_enabled** (bool): Enable CSRF protection
  - Defaults to `False` in development (`ENV=dev`), `True` otherwise
```

- **csrf_methods** (Set[str]): HTTP methods requiring CSRF protection
  - Default: `{'POST', 'PUT', 'PATCH', 'DELETE'}`

### Cookie Configuration

- **secure_cookie** (bool): Enable secure cookie attributes
  - Default: `True` (when `NEXTAUTH_URL` starts with https)

- **cookie_name** (str): Session token cookie name
  - Default: `"__Secure-next-auth.session-token"` (when secure_cookie is True)
  - Default: `"next-auth.session-token"` (when secure_cookie is False)
  - For Auth.js (latest), use: `"__Secure-authjs.session-token"` (when secure_cookie is True)

- **csrf_cookie_name** (str): CSRF token cookie name
  - Default: `"__Host-next-auth.csrf-token"` (when secure_cookie is True)
  - Default: `"next-auth.csrf-token"` (when secure_cookie is False)

>   - For Auth.js (latest), use: `"__Host-authjs.csrf-token"` (when secure_cookie is True)

Reviewer: These already use auth.js names in the latest version

Reviewer: Please check this again, the cookie names are already the correct authjs ones by default

> [!TIP]
> If you're using the latest version of Auth.js, here's the recommended configuration:
> ```python
> JWT = NextAuthJWT(
>     secret=os.environ["AUTHJS_SECRET"],
>     cookie_name="__Secure-authjs.session-token",
>     csrf_cookie_name="__Host-authjs.csrf-token",
> )

> > ```

Reviewer: The code still needs to be adapted to use AUTHJS_SECRET by default, but the cookie names should already be correct by default

- Uses v4-compatible cookie names
- Disables salt auto-appending

## Security Best Practices

1. **Environment Variables**: Always use environment variables for sensitive values:
   ```python
   JWT = NextAuthJWT(
       secret=os.getenv("NEXTAUTH_SECRET"),
   )
   ```

2. **HTTPS in Production**: Ensure `NEXTAUTH_URL` starts with `https://` in production to enable secure cookies

3. **CSRF Protection**: Keep CSRF protection enabled in production environments

## Examples

A [simple example](https://github.com/TCatshoek/fastapi-nextauth-jwt/tree/main/examples/simple) is available in the examples folder. It demonstrates:
- Using Next.js URL rewrites to route requests to FastAPI
- Basic JWT validation setup
- CSRF protection configuration

You can also place both the backend and frontend behind a reverse proxy like nginx, as long as cookies reach FastAPI.

## Environment Variables

- `NEXTAUTH_SECRET`: The secret key used for JWT operations (required)
- `NEXTAUTH_URL`: The URL of your application (affects secure cookie settings)
- `ENV`: Set to `"dev"` to disable CSRF protection in development

## Contributing

Contributions are welcome! Please feel free to submit a Pull Request.

> 

Reviewer: I'm very short on time lately, so I can't promise I have time to review PRs in a timely manner :(

> [![Star History Chart](https://api.star-history.com/svg?repos=TCatshoek/fastapi-nextauth-jwt&type=Date)](https://star-history.com/#TCatshoek/fastapi-nextauth-jwt&Date)

Reviewer: I'd personally leave this out as it is not really relevant
